### PR TITLE
Add log-source-app for diagnostic log cluster example

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
diagnostic log cluster implementation is missing

#### Change overview
Add a example app for diagnostic log cluster

#### Testing
Manually verified by running the app.

### CODE REVIEW

File: src/lib/support/BytesCircularBuffer.cpp [hunk 1]
```diff
@@ -111,6 +111,32 @@ CHIP_ERROR BytesCircularBuffer::Push(const ByteSpan & payload)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR BytesCircularBuffer::Push(const ByteSpan & payload1, const ByteSpan & payload2)
+{
+    size_t length = payload1.size() + payload2.size();
+    if (length > std::numeric_limits<SizeType>::max())
```
Comment (andy31415) [inline src/lib/support/BytesCircularBuffer.cpp:117]: nit: add {} for ifs here and below

**Review by andy31415 — APPROVED** [review]

File: examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <log-source-common/DiagnosticLogsCommandHandler.h>
+
+CHIP_ERROR DiagnosticLogsCommandHandler::PushLog(const chip::ByteSpan & payload)
```
Comment (Damian-Nordic) [inline examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.cpp:21]: Is there a reason it is implemented in the example app instead in the common `src/app` module? Even the logging redirection is platform-agnostic, so I'm wondering if most of this functionality couldn't be implemented in some generic (although configurable) way?

File: examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include <app-common/zap-generated/cluster-id.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandlerInterface.h>
+#include <lib/support/BytesCircularBuffer.h>
+
+#include <array>
+
+/**
+ * A reference implementation for an OTA Provider. Includes a method for providing a path to a local OTA file to serve.
+ */
```
Comment (tcarmelveilleux) [inline examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.h:30]: This has cut'n'paste of OTA comments

File: examples/log-source-app/linux/main.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+constexpr uint16_t kOptionFilepath = 'f';
+
+// Arbitrary BDX Transfer Params
+constexpr uint32_t kMaxBdxBlockSize                 = 1024;
+constexpr chip::System::Clock::Timeout kBdxTimeout  = chip::System::Clock::Seconds16(5 * 60); // Specification mandates >= 5 minutes
+constexpr chip::System::Clock::Timeout kBdxPollFreq = chip::System::Clock::Milliseconds32(500);
+
+// Global variables used for passing the CLI arguments to the DiagnosticLogsCommandHandler object
+const char * gOtaFilepath = nullptr;
+
```
Comment (tcarmelveilleux) [inline examples/log-source-app/linux/main.cpp:56]: This is cut'n'paste of OTA examples that are already using some temporary ways of doing BDX session management. Strongly recommend against copying, and certainly recommend against keeping OTA constants  in here

File: examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.cpp (hunk 2 above)
Comment (kghost) [reply examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.cpp:21]: Unlike other clusters, the logs source/sink doesn't fit the server/client model. The controller (aka client), can either by source or sink. So I'm not sure it is feasible to put it in `src/app/clusters`

File: examples/log-source-app/linux/main.cpp (hunk 4 above)
Comment (kghost) [reply examples/log-source-app/linux/main.cpp:56]: These are needed for log file transfer. But as long as it is not implemented yet, and these fields are not being used, I'll remove them now.

File: examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+class DiagnosticLogsCommandHandler : public chip::app::CommandHandlerInterface
+{
+public:
+    static constexpr const uint16_t kDiagnosticLogsEndpoint   = 0;
+    static constexpr const uint16_t kDiagnosticLogsBufferSize = 4 * 1024; // 4K internal memory to store text logs
```
[the remark this answers is not in the document]
Comment (kghost) [reply examples/log-source-app/log-source-common/DiagnosticLogsCommandHandler.h:35]: This is already the example part, it is expected to be copied into production code, I'm not sure there is any way to configure values in examples.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking, given this has had enough time for review. Comments still need to be resolved though.

**Review by woody-apple — APPROVED** [review]
Fast tracking, this has been available for review for > 3 days

**Pull request merged**